### PR TITLE
refactor: update handle function of cc_mounts

### PR DESCRIPTION
**Pull request by TheRealFalcon**
## Proposed Commit Message

```
refactor: update handle function of cc_mounts

The handle function of cc_mounts was hard to grok and had one of the
highest cyclomatic complexity scores in the codebase. Functionally,
the code should be unchanged.
```

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by TheRealFalcon** [comment]
cc @a-dubs

File: cloudinit/config/cc_mounts.py
Comment (a-dubs) [inline cloudinit/config/cc_mounts.py:140]: could we reword this description a little?

**Review by a-dubs — APPROVED** [review]
Changes look good! Left one little comment/nit but not going to block on it. 

Also I was wondering if it would be worth the time to add further unit tests or integration tests before merging to ensure this refactor didn't break anything?

**Comment by TheRealFalcon** [comment]
> Also I was wondering if it would be worth the time to add further unit tests or integration tests before merging to ensure this refactor didn't break anything?

The last two commits in https://github.com/canonical/cloud-init/pull/5459 added some coverage for things I knew I was changing here. Got any specific ideas?

**Comment by a-dubs** [comment]
@TheRealFalcon After reading back over all of the cc_mounts unit tests, I realize there is indeed coverage for all the functions you have created. My original thought was, "oh wow thats a lot of new helper functions, we should unit test those!!!". But I now see that the unit tests pretty thoroughly test the functions that rely on these helpers, so their functionality is pretty well tested, just not directly. So I think the effort of adding new unit tests is not worth the reward.

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Minor nits and I think a potentially significant question on `santize_mounts_configuration` to see if we are accidentally dropping items from mount config lines if they contain `None`.

File: cloudinit/config/cc_mounts.py
Comment (TheRealFalcon) [reply cloudinit/config/cc_mounts.py:472]: Those work too, but I think I prefer it as-is.

File: cloudinit/config/cc_mounts.py
Comment (TheRealFalcon) [reply cloudinit/config/cc_mounts.py:434]: Because we have some weird rules that would allow a mount declaration like this:
```
/dev/sda1 /mnt1
/dev/sda1 None
/dev/sda1 /mnt2
```

The `None` on line 2 means "disregard this line and any /dev/sda1 entries that came before it". The first reversal is to make it easy to know which lines to throw out that we come upon. If we previously saw a None having the same first entry, we know to throw it out.

Since we accumulated our act list in reverse order, the last reverse is to put the list back into it's original order (after we've removed the lines we want to remove).

File: cloudinit/config/cc_mounts.py
Comment (TheRealFalcon) [reply cloudinit/config/cc_mounts.py]: Ahh, good find. I reworked the code there a bit to align with what we had before and also updated a test case that should pass on main but fail pre-fixing this.

File: cloudinit/config/cc_mounts.py
Comment (TheRealFalcon) [reply cloudinit/config/cc_mounts.py]: Nice! I like it

File: cloudinit/config/cc_mounts.py
Comment (TheRealFalcon) [reply cloudinit/config/cc_mounts.py:549]: No actually. I could specify `ephemeral None` to tell cloud-init not to render the epehermal device. On `add_default_mounts_to_cfg`, it'll see that I have already defined it, and then not touch the mount list, then later remove it on the `remove_nonexistent_devices` call. If I switch the order, cloud-init will remove it from the list, then add the default back in. We'll get the default added to the mount list when we didn't want it.

Comment (blackboxsw) [reply cloudinit/config/cc_mounts.py:549]: Good, thanks for the correction/explanation there.

**Comment by TheRealFalcon** [comment]
@blackboxsw , this should be ready for re-review

File: cloudinit/config/cc_mounts.py
Comment (blackboxsw) [inline cloudinit/config/cc_mounts.py:379]: I wondered about defining a namedtuple here for readability and usability of the return value from this function but I don't think it's worth the minor overhead here and we only have one call-site to keep aligned with this return value, so not a lot of reuse potential there.

**Review by blackboxsw — APPROVED** [review]
LGTM. Thanks for the fixup of potentially redacted values and unittest additions. Validated before and after expected values in /etc/fstab mounts remain unchanged in integration testing and test coverage looks good.

This may have conflicts in merge with upstream/main to resolve.

File: cloudinit/config/cc_mounts.py
Comment (TheRealFalcon) [reply cloudinit/config/cc_mounts.py:379]: Agreed. Since it's just one site it doesn't seem like it would add much value.

**Pull request merged**
